src/metadata.rs

```diff
@@ -165,6 +171,28 @@ impl Metadata {
         }
         string
     }
+
+    /// Derive a mortal period
+    pub(crate) fn derive_mortal_period(&self) -> Result<u64, MetadataError> {
```

Comment:
I was thinking a little bit about having concrete types `u32 and u64` here which might not work for all runtimes ideally we would read the types from the runtime but it seems quite complicated to get it to work and `pallet-babe` is not implemented for `substrate-subxt`.

Is it fair to assume that `BABE` is used for all runtimes?

However, for example the `runtime type parameter` was passed in here we could do:
```rust
    /// Derive a mortal period
    pub(crate) fn derive_mortal_period<T: Runtime>(&self) -> Result<impl AtLeast32BitUnsigned, MetadataError> {
        let block_hash_count = self
            .module("System")
            .and_then(|sys| sys.constant("BlockHashCount"))
            .and_then(|count| count.value::<<T as System>::BlockNumber>())?;

        let expected_block_time = self
            .module("Babe")
            .and_then(|babe| babe.constant("ExpectedBlockTime"))
            // babe frame trait doesn't exist yet....
            .and_then(|e| e.value::<<T as Babe>::ExpectedBlockTime>())?;

        // maybe convert by try_from or something.
        match expected_block_time {
            0 => Err(MetadataError::ZeroValue("Babe::ExpectedBlockTime")),
            expected_block_time => {
                Ok((BASELINE_MORTAL_PERIOD / expected_block_time)
                    .next_power_of_two()
                    .min(block_hash_count.next_power_of_two()))
            }
        }
    }
```

//cc @ascjones

Reply:
As we speak I am working on upgrading substrate metadata to use `scale-info` for full type information and once we have that we should be able to do runtime specific builds for `subxt`